ACROSS API: Add BurstCube TOO CRUD API

jak574 opened the pull request:
### ACROSS API: Add BurstCube TOO CRUD API

### Description

PR adds BurstCubeTOO GET / POST / DELETE / PUT endpoints for submission, editing, fetching and deleting of BurstCube TOOs. This also provides the backend database and constraint checking for incoming TOOs. Broadly this PR addresses the issues listed in Issue #1976.


### Reviewers
Leo Singer
Samuel Wyatt

### Changes
This adds the following classes:

`BurstCubeTOO` - class to handle CRUD for BurstCubeTOO events which are made available through the API using standard `GET`, `POST`, `PUT` and `DELETE`. `POST` (i.e. submitting a new TOO) initiates checks on feasibility of the particular trigger. If coordinates are given, then probability of being in the BurstCube FOV are performed using the `BurstCubeFOV` class. In addition the TOO is checked for age and if BurstCube was inside the South Atlantic Anomaly 

`BurstCubeTOORequests` - a class to handle the fetching of multiple TOO requests based on date range, number limit, or most recent number of days. 

The adds the following endpoint to the api interface: `/burstcube/too`. For `PUT`, `POST` and `DELETE` authentication is required. For `GET` no authentication is required. `PUT`, `DELETE` and optionally `GET` pass the TOO `id` which is generated and returned with a `POST` using the form `/burstcube/too/{id}`. For `GET` if `id` is not passed, then `begin` and `end` parameters can be passed to specify date ranges, number limits and length of days to fetch using the `BurstCubeTOO` class.

### Acceptance Criteria
Passing code review and CI checks.

### Related Issue(s)
Fixes #1976
#1899

### Testing
~~pytest tests are included to test CRUD actions and a to validate some rejection criteria (double TOO submission, trigger > 48 hours old).~~
pytests will be added in a later PR

lpsinger commented on app.arc:
```diff
@@ -49,6 +49,25 @@ acrossapi_tle
+burstcube_too
+  id *String
+  created_on **String
+  ra Number
+  dec Number
+  error_radius Number
+  created_by String
+  created_on String
+  modified_by String
+  modified_on String
+  trigger_time String
+  trigger_info String
+  exposure Number
+  offset Number
+  reject_reason String
+  status String
+  too_info String
```
Invalid syntax. Only define the types for the partition key and sort key.

```suggestion
```

lpsinger commented on python/across_api/base/constraints.py:
```diff
@@ -41,7 +41,7 @@ def get_slice(time: Time, ephem: EphemBase) -> slice:
         # Check that the time range is within the ephemeris range, as above.
         assert (
-            time[0] >= ephem.begin and time[-1] <= ephem.end
+            time[0].unix >= ephem.begin.unix and time[-1].unix <= ephem.end.unix
```
Is it important to convert to UNIX time? Does astropy.time.Time not support comparison between Time objects with differing scales and representations?

jak574 replied:
Actually something more fundamental may be happening here. Consider this code:

```python
from time import time
from astropy.time import Time
from datetime import datetime

t = time()
dt = datetime.utcfromtimestamp(t)

assert Time(t, format='unix').unix == Time(dt.utcfromtimestamp(t)).unix, "This should work"
assert Time(t, format='unix') == Time(dt.utcfromtimestamp(t)), "This won't work"
```
It's not clear to me why, but comparing `Time` constructed from a `datetime` and a unix time are not equal.

jak574 replied:
Consider this code:

```python
from time import time
from astropy.time import Time
from datetime import datetime

t = time()
dt = datetime.utcfromtimestamp(t)
assert Time(t, format='unix').unix == Time(dt).unix, "This works"
assert Time(t, format='unix') == Time(dt), "This won't work"
```
It fails at the 2nd assertion, despite the two sides of the assertion being generated from the same timestamp.

jak574 replied:
Looks like astropy has an `isclose` method in `Time` to deal with this sort of thing, but it's not obvious to me how you'd do `>=` with it.
```python
assert Time(t, format='unix').isclose(Time(dt), 1e-6 * u.s)
```

lpsinger replied:
Subsequent to this line, you call `ephindex` which computes an integer index. What you really want to guard against is that this index is within bounds. If it's out of bounds, will you get an IndexError, or will it silently do something weird? If the former, you may be able to do without this assertion.

jak574 replied:
It will silently do something weird. Best to keep in this assertion, which should never be triggered unless a bug is introduced.

lpsinger replied:
In that case, I suggest lowering this assertion down into `ephindex` and asserting that the index is within bounds. That kind of assertion will not be subject to floating point imprecision.

As written, this assertion assumes that all of the times under comparison have the same scale, which may be true right now but is a latent bug.

jak574 replied:
Sure, makes sense to do it there. Fixed in [ff24853](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/ff24853f09067b18e3910804aa9d768f502d822d).

lpsinger commented on python/across_api/base/schema.py:
```diff
@@ -22,13 +25,45 @@
+]
+
+
+# Pydantic type for a Astropy Time  in seconds
```
```suggestion
# Pydantic type for a Astropy Time in seconds
```

lpsinger commented on python/across_api/burstcube/schema.py:
```diff
@@ -0,0 +1,239 @@
+    """
+    Enumeration class representing the status of a Target of Opportunity (TOO) request.
+
+    Attributes:
+    requested : str
+        The TOO request has been submitted.
+    rejected : str
+        The TOO request has been rejected.
+    declined : str
+        The TOO request has been declined.
+    approved : str
+        The TOO request has been approved.
+    executed : str
+        The TOO request has been executed.
+    other : str
+        The TOO request has a status other than the predefined ones.
```
The quality of the contextual help in code editors may be better if you make these descriptions the docstrings of their respective enum attributes rather than of the class itself, like this:

```python
    requests = "Requested"
    """The TOO request has been submitted."""
```

jak574 replied:
I tried this and it didn't work in VSCode. The docstrings added after the Enum values are not displayed in tooltip hovers. However, the docstring for the whole class is shown. 

This is what it looks like with the docstring:

<img width="249" alt="Screenshot 2024-02-26 at 14 23 04" src="https://github.com/nasa-gcn/gcn.nasa.gov/assets/15330708/746692c8-edce-462d-aa23-bdbe6643f808">

without:

<img width="297" alt="Screenshot 2024-02-26 at 14 27 58" src="https://github.com/nasa-gcn/gcn.nasa.gov/assets/15330708/692ab747-6066-4f80-90ca-7c340004bf07">

This is what hovering over the enum looks like in both cases:

<img width="327" alt="Screenshot 2024-02-26 at 14 39 43" src="https://github.com/nasa-gcn/gcn.nasa.gov/assets/15330708/7cf97622-3bec-424b-8eb3-b125b5b8e7e3">

lpsinger replied:
Oh well. Sphinx renders docstrings for member variables. Thanks for checking.

lpsinger reviewed (changes requested):
DynamoDB has a per-item size limit of 400 KB. Will all HEALPix localizations fit inside that limit? Even if it does, we may want to store the HEALPix localizations outside of the database in an S3 bucket in order to avoid being billed for excess read request units when querying the table.

We already have a writable S3 bucket that is provisioned by Architect.

jak574 commented on python/across_api/burstcube/requests.py:
Mostly because I tend to not like having more than one class per file. That said, these two things are considated into a single endpoint now, so it makes sense to have them into in the same file. I'll do that.

jak574 replied:
Fixed by [b97a247](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/b97a247576afd92e70ecae00ec8ef4504b8a16e7)

jak574 commented on python/across_api/base/api.py:
```diff
@@ -19,3 +23,167 @@
+            ge=0,
+            lt=360,
+            title="RA (J2000)",
+            description="Right Ascenscion in J2000 coordinates and units of decimal degrees.",
+        ),
+    ] = None,
+    dec: Annotated[
+        Optional[float],
+        Query(
+            ge=-90,
+            le=90,
+            title="Dec (J2000)",
+            description="Declination in J2000 coordinates in units of decimal degrees.",
+        ),
+    ] = None,
+) -> Optional[dict]:
+    if ra is None or dec is None:
+        return {"ra": None, "dec": None}
+    return {"ra": ra * u.deg, "dec": dec * u.deg}
+
+
+OptionalRaDecDep = Annotated[dict, Depends(optional_ra_dec)]
+
+
+IdDep = Annotated[str, Path(description="TOO ID string")]
+
+
+async def trigger_time(
+    trigger_time: Annotated[
+        datetime,
+        Query(
+            title="Trigger Time",
+            description="Time of trigger in UTC or ISO format.",
+        ),
+    ],
+) -> Optional[datetime]:
+    return Time(trigger_time)
+
+
+TriggerTimeDep = Annotated[datetime, Depends(trigger_time)]
```
Good point. I'll do that.

jak574 replied:
Fixed in [439d0b1](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/439d0b1f64a7cbe9bac73f47e1f13518801bf127)

lpsinger commented on python/across_api/base/api.py:
```diff
@@ -19,3 +23,167 @@
     root_path="/labs/api/v1",
 )
+
+# FastAPI Depends definitions
```
Remove tautological comment.

```suggestion
```

lpsinger commented on python/across_api/base/api.py:
```diff
@@ -19,3 +23,167 @@
+
+
+# Depends functions for FastAPI calls.
+async def optional_length(
```
Rename to duration.

jak574 replied:
Fixed in [6ef39aa](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/6ef39aa41dc8bc60b462133ef6dbc16d2ee10f42)

lpsinger commented on python/across_api/base/api.py:
```diff
@@ -19,3 +23,167 @@
+    ] = None,
+) -> Optional[u.Quantity]:
+    """
+    Helper function to convert begin and end to datetime objects.
```
Copy-paste error in docstring.

jak574 replied:
Fixed in 121d3bf

lpsinger commented on python/across_api/base/api.py:
```diff
@@ -19,3 +23,167 @@
+
+# Depends functions for FastAPI calls.
+async def optional_daterange(
+    begin: Annotated[
+        Optional[datetime],
+        Query(
+            description="Start time of period to be calculated.",
+            title="Begin",
+        ),
+    ] = None,
+    end: Annotated[
+        Optional[datetime],
+        Query(
+            description="Start time of period to be calculated.",
+            title="End",
+        ),
+    ] = None,
```
astropy.time.Time can convert from many string formats in addition to datetime objects. Consider broadening types to anything that astropy.time.Time's constructor can accept.

jak574 replied:
This is about interaction with the API endpoints, which always spits out a datetime, so it's not really necessary to support other time formats for this code.

lpsinger replied:
I thought that these dependency functions preprocess parameters received from the request. Isn't that so?

Is this in the same category as #2004, something for a future PR?

jak574 replied:
The depends functions are there to create the API endpoint parameters, and drive what ends up in the SwaggerUI interface and the OpenAPI.json specification. OpenAPI supports dates and datetimes in format defined by [RFC 3339, section 5.6](https://tools.ietf.org/html/rfc3339#section-5.6) only. FastAPI converts these into a datetime (naive or not naive depending on whether a timezone is specified in the input). 

The depend function has two main uses:

1. It specifies a standardized parameter name and description for a parameter. By having a function that does this rather than just defining inline, you encourage consistency across the API and code reuse. After all, we're going to end up requesting `begin` and `end` a lot in across many missions. 

2. It converts what FastAPI spits out (in this case a `datetime`) to what we want (an astropy `Time`). 

I believe that our OpenAPI specification should stick with the basic definitions supported, and not use loosely defined strings for simple objects like dates and datetimes. 

Again, this is a design decision, and one we can revisit at a later date. I think it's out of scope here to change it without a detailed look at the pros and cons of how this is done involving the whole team.

lpsinger commented on python/across_api/base/api.py:
```diff
@@ -19,3 +23,167 @@
+            description="Error radius in degrees.",
+        ),
+    ] = None,
+) -> Optional[float]:
```
Incorrect type.

```suggestion
) -> Optional[u.Quantity['angle']]:
```

jak574 replied:
Fixed in [1a908cf](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/1a908cfc67a4771903e4aac26de1c5622bb97bd3). I'll note that mypy doesn't like `u.Quantity['angle']` so I used `u.Quantity[u.deg]`.

lpsinger commented on python/across_api/base/api.py:
```diff
@@ -19,3 +23,167 @@
+            description="Exposure time in seconds.",
+        ),
+    ] = 200,
+) -> u.Quantity:
```
```suggestion
) -> u.Quantity['time']:
```

jak574 replied:
Fixed in 121d3bf

lpsinger commented on python/across_api/base/api.py:
```diff
@@ -19,3 +23,167 @@
+) -> Optional[dict]:
+    if ra is None or dec is None:
+        return {"ra": None, "dec": None}
+    return {"ra": ra * u.deg, "dec": dec * u.deg}
```
Would be more useful to return an `astropy.coordinates.SkyCoord` instance.

jak574 replied:
This is going to lead to a bunch of unnecessary refactoring of the code, so let's address it in a later PR.

lpsinger replied:
See #2006.

lpsinger commented on python/across_api/base/api.py:
```diff
@@ -19,3 +23,167 @@
+            description="Time of trigger in UTC or ISO format.",
+        ),
+    ],
+) -> Optional[datetime]:
```
```suggestion
) -> Optional[Time]:
```

lpsinger commented on python/across_api/base/ephem.py:
```diff
@@ -86,6 +86,11 @@ def ephindex(self, t: Time) -> int:
         -------
             The index of the nearest time in the ephemeris.
         """
+        # Check that the time is within the ephemeris range.
+        assert (
+            t.jd >= self.begin.jd and t.jd <= self.end.jd
+        ), "Time outside of ephemeris of range"
```
This assertion is still incorrect. Please assert that the integer index computed in the line below is in bounds.

jak574 replied:
This is how the integer is calculated:

```int(np.argmin(np.abs((self.timestamp.datetime - t.datetime))))```

If `t` is outside of the scope of the Ephemeris, it will just return `0` or the maximum index value. So asserting i < 0 or i > i_max will *never* be triggered.

lpsinger replied:
I don't fully understand what `argmin` does here, but in general if you are converting from a floating point value to an index into a bounded array, I would do it like this:

```python
i = int((t - t_min) // dt)
assert i >= 0 && i < i_max
```

If you write it instead like this:

```python
assert t >= t_min && t < t_max
i = int((t - t_min) // dt)
```

then it is likely that the assertion is incorrect.

jak574 replied:
The nuance of using `np.argmin` here so that if you have a time of 1.999999s, and are searching a time series that has 1 second resolution, it finds the closest value, so would return the index for 2s. The way you suggest would return the index for 1s.

jak574 replied:
This... I HOPE... satisfies both of us. [e1ea448](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/e1ea448ed4b989433ddc6f02a242c9ee9c639a65)

lpsinger commented on python/across_api/base/api.py:
Wherever we have dependencies that convert floats to Astropy quantities, consider also accepting strings so that we can take advantage of Astropy's nice parsing of things like `'1.5 arcmin'`.

jak574 replied:
Angles in a JSON API should be in a fixed unit, rather than a freeform string. Support for people using string definitions for angles like this belongs in the API client.

lpsinger replied:
I disagree; this would be great to have at the API level. If implemented at the API level, then this functionality is available for all languages.

jak574 replied:
This is an API design change, so an out of scope change for a PR. Let's revisit this at a later date.

lpsinger replied:
Fair enough. I've created #2004 to track that.

lpsinger reviewed (changes requested):
- [ ] I'm not crazy about adding moto as a dependency or mocking DynamoDB for unit tests. We have a mostly complete local implementation of the AWS API including a local, embedded instance of DynamoDB: `arc sandbox`. And if we really need to test things in an even more isolated way that `arc sandbox`, then I would prefer to use pytest-dynamodb which also embeds a real DynamoDB database. We use that in the unit tests for dynamodb-autoincrement.
- [ ] I had thought that we had agreed to make AWS API calls asynchronously using aioboto3 rather than synchronously using boto3.

jak574 commented:
> * [ ]  I had thought that we had agreed to make AWS API calls asynchronously using aioboto3 rather than synchronously using boto3.

I'm using architect to access AWS, not boto3. This is what I used for the TLE database, so I'm using it here to be consistent.

I'm down with using aioboto3, but it makes the code inconsistent and I don't want to make changes to code that isn't covered by this PR unless absolutely necessary.

I'll do a PR to change over to aioboto3 after this PR is merged.

lpsinger commented:
> I'll do a PR to change over to aioboto3 after this PR is merged.

Great, let's defer that for later.

jak574 commented:
> * [ ]  I'm not crazy about adding moto as a dependency or mocking DynamoDB for unit tests. We have a mostly complete local implementation of the AWS API including a local, embedded instance of DynamoDB: `arc sandbox`. And if we really need to test things in an even more isolated way that `arc sandbox`, then I would prefer to use pytest-dynamodb which also embeds a real DynamoDB database. We use that in the unit tests for dynamodb-autoincrement.

I removed the moto dependency and replaced it with pytest-dynamodb. As discussed in the Pulse earlier, an Architect based set of pytests can be added in a future PR.

Note that to get pytests running in the workflow, I had to update the workflow, including copying the dynamodb-install action from your [nasa-gcn/python-dynamodb-autoincrement ](https://github.com/nasa-gcn/python-dynamodb-autoincrement) repo. I'm sure there's a better way to handle that, but for now, it works and the tests pass.

lpsinger commented:
> > * [ ]   I had thought that we had agreed to make AWS API calls asynchronously using aioboto3 rather than synchronously using boto3.
> 
> I'm using `architect-functions` to access AWS, not boto3. This is what I used for the TLE database, so I'm using it here to be consistent.
> 
> I'm down with using aioboto3, but it makes the code inconsistent and I don't want to make changes to code that isn't covered by this PR unless absolutely necessary.
> 
> I'll do a PR to change over to aioboto3 after this PR is merged.

See #2005.

lpsinger commented on python/across_api/burstcube/schema.py:
```diff
@@ -0,0 +1,219 @@
+
+    model_config = ConfigDict(extra="allow")
+
+    @model_validator(mode="before")
+    def convert_json_string_to_dict(cls, data):
+        if isinstance(data, str):
+            return json.loads(data)
+        return data
```
Why are we converting it to and from a string? Why not just store it as a dictionary?

jak574 replied:
It needs to be converted to a JSON string so it can be written to DynamoDB under the key `trigger_info`. When it gets read back it gets converted back into a dict.

lpsinger replied:
Ah, I see the confusion. DynamoDB records are JSON documents. It does not need to be converted to a string.

jak574 replied:
So I figured out that the reason I was doing this was due to a problem with how I was writing TOOs to DynamoDB. The main issue being the need to have numbers be `Decimal`. I had been handling this with a conversion Schema called `BurstCubeTOOModel`, but this broke when adding `BurstCubeTriggerInfo` as a sub class, and my solution was to convert it to a JSON string. This worked, but as you point out, wasn't necessary.

So I got rid of `BurstCubeTOOModel` and now just convert `BurstCubeTOOSchema` into a dict with all `float`s converted to `Decimal`. This works, allowed me to delete `BurstCubeTOOModel` and a bunch of other code, and is much cleaner. 

Fixed by [2c9c6f3](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/2c9c6f3059423867cdfecb464ae0cdc06a766622) and [b02735d](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/b02735d5c2ae03ff2381fc41dfb1c8b67f323e9d). 

I need to push an update to the tests as they are now failing.

jak574 replied:
So I had to put this code back in. [ce858aa](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/ce858aabffed8445410dfc1091a41e82fd05e4f1)

I had forgotten part of why this exists. It's due to the complexities of HTTP POST. POST does not allow you to submit files and JSON data, so you have to submit any JSON data as a file, which means it's encoded as a JSON `str`, not as a `dict`. This means that it fails `model_validate` unless you put in a validator (like the removed code) that parses JSON `str` into a `dict`. 

This is either a limitation of FastAPI or there is some esoteric way around it I cannot figure out. I found one discussion that suggests the former: https://github.com/tiangolo/fastapi/discussions/9868

This highlights the need for tests which submit through the API running in a sandbox environment, rather than just testing at the python class level, because a change that breaks the API interface right now is not going to be caught.

lpsinger commented on python/across_api/burstcube/schema.py:
```diff
@@ -0,0 +1,219 @@
+    status: str = "Requested"
+    too_info: str = ""
+
+    @computed_field  # type: ignore
+    @property
+    def id(self) -> str:
+        return hashlib.md5(
+            f"{self.trigger_time}{self.exposure}{self.offset}".encode()
+        ).hexdigest()
```
Remind me again, why do models need to be hashable?

jak574 replied:
This code is just creating a unique ID key for each TOO based on the three unique required parameters (and no more).

lpsinger replied:
Ah, so the reason is that you decided to use these to form the partition key for this table?

lpsinger commented:
Blocked by https://github.com/nasa-gcn/gcn.nasa.gov/issues/2009.

lpsinger commented on python/across_api/base/database.py:
```diff
@@ -0,0 +1,14 @@
+import os
+
+
+def dynamodb_table(tablename) -> Any:
+    """If running in Architect, return tables.table, else return boto3 dynamodb
+    table. This enables the use of moto to mock the dynamodb table in tests."""
+    if os.environ.get("ARC_ENV") is not None:
+        return arc.tables.table(tablename)
+    else:
+        session = boto3.Session()
+        return session.resource("dynamodb", region_name="us-east-1").Table(tablename)
```
Remove; you only added this to support mock testing.

jak574 replied:
It's gone

lpsinger commented on python/across_api/burstcube/api.py:
```diff
@@ -1,3 +1,184 @@
+    # Type hint for the file object
+    file: Union[gzip.GzipFile, BinaryIO]
+
+    # If the file is a gzip file, open it with gzip.open, otherwise just
+    # pass along the filehandle
+    if healpix_file.content_type == "application/x-gzip":
+        file = gzip.open(healpix_file.file, "rb")
+    else:
+        file = healpix_file.file
```
Is this necessary? I thought that `fits.open` would transparently decompress the file for you.

lpsinger replied:
According to https://docs.astropy.org/en/stable/io/fits/#working-with-compressed-files:

> The [open()](https://docs.astropy.org/en/stable/io/fits/api/files.html#astropy.io.fits.open) function will seamlessly open FITS files that have been compressed with gzip, bzip2 or pkzip. Note that in this context we are talking about a FITS file that has been compressed with one of these utilities (e.g., a .fits.gz file).

It may be that the gzip detection is based on filename and so it doesn't work with file-like objects. Would you please check if that's the case?

jak574 replied:
As far as I can tell looking through the code and experimenting with it, the compression detection does not work with a `BytesIO` object. It will work if you do pass it a regular `open` object of a gzipped file. Seems like maybe a bug in astropy? I can file an Issue.

lpsinger replied:
Thanks! Please add a FIXME comment with a link to the issue.

There might already be an issue for this.

jak574 replied:
I tried to find an existing issue, but was unsuccessful, so filed one.

https://github.com/astropy/astropy/issues/16171

Added FIXME in 134a7b0.

lpsinger commented on python/across_api/burstcube/api.py:
```diff
@@ -1,3 +1,184 @@
+    """
+    # Construct the TOO object.
+    too = BurstCubeTOO(
+        username=credential["sub"],
```
This field should be called `sub`, not `username`. OIDC subject claims are not usernames.

jak574 replied:
Fixed

lpsinger commented on python/across_api/burstcube/api.py:
```diff
@@ -1,3 +1,184 @@
+    "/burstcube/too",
+    status_code=status.HTTP_201_CREATED,
+    dependencies=[
+        Security(scope_authorize, scopes=["gcn.nasa.gov/kafka-public-consumer"])
```
This scope will authorize every user who is logged in. You probably want to create a new scope for this.

jak574 replied:
Yes, this is what Issue #1901 is about. I don't know what that scope will be yet, so just set the default.

lpsinger replied:
Alright, let's create that new scope. Please leave comments on what you want to call the scope at https://github.com/nasa-gcn/gcn.nasa.gov/issues/1901#issuecomment-1980819293.

lpsinger commented on python/requirements.in:
```diff
@@ -6,6 +6,7 @@ email-validator
 fastapi
 healpy
 mangum
+python-multipart
```
I don't see any new imports related to this package.

jak574 replied:
It's required for FastAPI to support multi-part POST, e.g.. uploading of a file.

lpsinger replied:
```suggestion
python-multipart  # Required for file uploads in FastAPI; see https://fastapi.tiangolo.com/tutorial/request-files/
```

lpsinger commented on python/across_api/burstcube/toorequest.py:
```diff
@@ -0,0 +1,437 @@
+            raise HTTPException(404, "BurstCubeTOO not found.")
+
+        # Reconstruct the TOO as it exists in the database
+        print(response["Item"])
```
Remove print statement.

jak574 replied:
Done.

lpsinger commented on python/across_api/burstcube/toorequest.py:
```diff
@@ -0,0 +1,437 @@
+            response = self.table.get_item(Key={"id": self.id})
+            if response.get("Item"):
+                raise HTTPException(409, "BurstCubeTOO already exists.")
+        except botocore.exceptions.ClientError:
+            pass
```
This is a very broad `except:` statement. What errors are you trying to handle?

jak574 replied:
Yeah, not sure honestly. I just removed them, as I don't consistently try to trap exceptions with table read/writes. a474f4a5

lpsinger commented on python/across_api/burstcube/toorequest.py:
```diff
@@ -0,0 +1,437 @@
+    """
+    Class to fetch multiple BurstCubeTOO requests, based on various filters.
+
+    Note that the filtering right now is based on DynamoDB scan, which is not
+    very efficient. This should be replaced with a query at some point.
```
Indeed, that's both inefficient and expensive. Please make the time the sort key of the table and use the `<`, `<=`, `>`, `>=`, or `BETWEEN` operators in the [`KeyConditionExpression`](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html#DDB-Query-request-KeyConditionExpression).

jak574 replied:
I implemented this using a GSI (see checking [7e476a7](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/7e476a750d1bb78f9273772c772b5068b4c7924a))

lpsinger replied:
Do you need the dummy partition key `gsipk`? Could you instead use the id as the partition key for the secondary index?

jak574 replied:
AFAIK you have to give a partition key value for a query where you're filtering on the sort key. You can only do `between` on a sort key, so I have to define a GSI PK that is constant across the whole database. If there's another way to do this, let me know.

lpsinger commented on python/across_api/burstcube/toorequest.py:
```diff
@@ -0,0 +1,437 @@
+        """
+        Get a list of BurstCubeTOO requests
+
+        Returns
+        -------
+        bool
+            Did this work? True | False
```
Why return `False` on failure? Why not raise an exception?

jak574 replied:
To be fair, it won't ever return `False`. I removed that line.

lpsinger commented on python/across_api/burstcube/api.py:
```diff
@@ -1,3 +1,184 @@
+        exposure=exposure,
+        offset=offset,
+    )
+    # If a HEALpix file was uploaded, open it and set the healpix_loc
+    # and healpix_scheme attributes.
+    if healpix_file is not None:
+        too.healpix_loc, too.healpix_scheme = read_healpix_file(healpix_file)
```
We should store HEALPix files in S3, not DynamoDB. HEALPix files will exceed the per-item size DynamoDB item size limit and this will fail.

jak574 replied:
Verified, we're not storing healpix map in the database.

lpsinger commented:
@jak574, this basically looks good to me, but you may want @swyatt7 to provide a closer inspection.

swyatt7 reviewed (approved):
lgtm
